Test/paulmck/rcu.hpp

```diff
@@ -3,45 +3,26 @@
 #include <cstddef>
 #include <memory>
 #include <mutex>
+#include <tuple>
 #include <utility>
 
 // Derived-type approach.  All RCU-protected data structures using this
 // approach must derive from std::rcu_obj_base, which in turn derives
 // from std::rcu_head.  No idea what happens in case of multiple inheritance.
 
 namespace std {
-    template<typename T, typename D = default_delete<T>, bool E = is_empty<D>::value>
-    class rcu_obj_base: private rcu_head {
-        D deleter;
+    template<typename T, typename D = default_delete<T>>
+    class rcu_obj_base: private rcu_head, private std::tuple<D> {
```

Comment:
I didn't find anything in the standard I am afraid. The reason why I am suggesting this is that libstdc++'s `std::unique_ptr` implementation stores its pointer and deleter like this:

    typedef std::tuple<typename _Pointer::type, _Dp>  __tuple_type;
    __tuple_type                                      _M_t;

libstdc++ does not actually provide that guarantee. It will disable EBO in if the argument is a final class or a tuple. According to their comments,

    // Using EBO for elements that are tuples causes ambiguous base errors.